linuxkm/lkcapi_glue.c

```diff
@@ -291,30 +289,42 @@ static int km_AesCbcEncrypt(struct skcipher_request *req)
         return err;
     }
 
-    err = wc_AesSetIV(ctx->aes_encrypt, walk.iv);
+    /* Copy the cipher state to mitigate races on Aes.reg and Aes.tmp. */
+    aes_copy = (struct Aes *)malloc(sizeof(Aes));
+    if (aes_copy == NULL)
+        return -ENOMEM;
+    XMEMCPY(aes_copy, ctx->aes_encrypt, sizeof(Aes));
+
+    err = wc_AesSetIV(aes_copy, walk.iv);
 
     if (unlikely(err)) {
         pr_err("%s: wc_AesSetIV failed: %d\n",
                crypto_tfm_alg_driver_name(crypto_skcipher_tfm(tfm)), err);
-        return -EINVAL;
+        err = -EINVAL;
+        goto out;
     }
 
     while ((nbytes = walk.nbytes) != 0) {
-        err = wc_AesCbcEncrypt(ctx->aes_encrypt, walk.dst.virt.addr,
+        err = wc_AesCbcEncrypt(aes_copy, walk.dst.virt.addr,
                                walk.src.virt.addr, nbytes & (~(WC_AES_BLOCK_SIZE - 1)));
 
         if (unlikely(err)) {
             pr_err("%s: wc_AesCbcEncrypt failed for %u bytes: %d\n",
                    crypto_tfm_alg_driver_name(crypto_skcipher_tfm(tfm)), nbytes, err);
-            return -EINVAL;
+            err = -EINVAL;
+            goto out;
         }
 
         nbytes &= WC_AES_BLOCK_SIZE - 1;
         err = skcipher_walk_done(&walk, nbytes);
     }
 
     /* copy iv from wolfCrypt back to walk.iv */
-    XMEMCPY(walk.iv, ctx->aes_encrypt->reg, WC_AES_BLOCK_SIZE);
+    XMEMCPY(walk.iv, aes_copy->reg, WC_AES_BLOCK_SIZE);
```

Comment:
Just double checking: the way this works is `walk->iv = req->iv;` is set in `skcipher_walk_skcipher()` in `skcipher_walk_virt()`, which is how the updated state persists btwn encrypt/decrypt requests

Reply:
Yeah -- the in-tree code can and does use `walk.iv` directly, passed as an arg to the encrypt/decrypt implementations (which, notably, saves them the inefficiency of cloning the key schedules).

And btw, the main thing we're accommodating here by copying the updated IV back to `walk->iv` is to assure the crypto manager (the self-test harness) sees the expected final IV, which it compares to known-good values after we return.  In theory it would allow continuations through multiple calls, with multiple walks, with IV continuity, but I don't know if there's kernel code currently that does that.  It must and would work, in any case.